feature/리뷰 등록 및 리뷰 목록 불러오기 기능 추가#24

Jeonminje opened the pull request:
## MovieController
- 리뷰 등록 기능이 추가되었습니다.
- 리뷰 목록 : json형식으로 반환해주었으며, 리뷰의 총 갯수, 총 페이지 개수, 현재 페이지, 현재 페이지의 data를 리턴해주었습니다.
```java
@PostMapping("/review")
    @ResponseStatus(HttpStatus.OK)
    public String saveReview(@RequestBody Review review){
        movieService.saveReview(review);
        return SUCCESS_SAVE_REVIEW.getMessage();
    }

    @GetMapping("/review/list")
    @ResponseStatus(HttpStatus.OK)
    public String getReviewList(@RequestParam(defaultValue = "1") Integer currentPageNum){

        return movieService.getReviewList(currentPageNum).toString();
    }
```

## Review, ReviewVO
- 리뷰를 받고 db에 저장하기위한 Review 클래스가 추가되었습니다.
- review 테이블, member 테이블을 join 후 데이터를 담아줄 ReviewVO 클래스가 추가되었습니다.

## ReviewMapper
- DB의 review 테이블에 접근하기 위한 매퍼를 추가했습니다.
- review의 총 갯수를 반환해주는 getTotalRowCount() 함수가 추가되었습니다.
- getReviewList() 로 join한 행들을 List로 반환해 주었습니다.
```sql
    <insert id="save" useGeneratedKeys="true" keyProperty="reviewId">
        insert into review(movie_id, member_id, content, rating, updated_at, created_at)
        values (#{movieId}, #{memberId}, #{content},#{rating}, current_timestamp , current_timestamp)
    </insert>

    <select id="getTotalRowCount" resultType="int">
        select count (*)
        from review
    </select>

    <select id="getReviewList" resultType="ReviewVO">
        select r.id as reviewId, movie_id, login_id, r.member_id as member_id, nickname, content, rating
        from review as r left join member as m
        on r.member_id = m.id
        limit #{startIdx}, 10
    </select>
```

## SessionManager
- 기존 부적절한 함수 이름을 수정해 주었습니다.

## MovieService
- 리뷰 저장
```java
public void saveReview(Review reviewDto) {
        Long memberId = (Long) httpSession.getAttribute(SessionConst.MEMBER_ID);

        Review review = Review.builder()
                .movieId(reviewDto.getMovieId())
                .memberId(memberId)
                .content(reviewDto.getContent())
                .rating(reviewDto.getRating()).build();
        reviewMapper.save(review);
    }
```
로그인 되어있는 상태이기 때문에 세션에서 memberId 값을 꺼내 review 객체를 생성 해 주었습니다.

- 리뷰 불러오기
```java
 public JSONObject getReviewList(Integer currentPageNum){

        int total = reviewMapper.getTotalRowCount();
        int reviewPerPage = 10;
        int totalPageNum = (int) Math.ceil((double) total/reviewPerPage);
        int startIdx = (currentPageNum - 1) * reviewPerPage;

        JSONObject jsonObject = new JSONObject();
        jsonObject.put("total", total);
        jsonObject.put("totalPageNum", totalPageNum);
        jsonObject.put("currentPage", currentPageNum);
        jsonObject.put("data", reviewMapper.getReviewList(startIdx));

        return jsonObject;
    }
```
지정된 페이지의 데이터와 리뷰관련 정보를 담아주었습니다.

Jeonminje committed:
리뷰 등록 및 리뷰 목록 불러오기 기능 추가

Jeonminje committed:
Merge branch 'loginInterceptor#20' into Movie-review#24

# Conflicts:
#	src/main/java/study/movieservice/controller/aop/LoginCheckAop.java

nahokyun commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -0,0 +1,45 @@
+        int reviewPerPage = 10;
+        int totalPageNum = (int) Math.ceil((double) total/reviewPerPage);
+        int startIdx = (currentPageNum - 1) * reviewPerPage;
+
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("total", total);
+        jsonObject.put("totalPageNum", totalPageNum);
+        jsonObject.put("currentPage", currentPageNum);
+        jsonObject.put("data", reviewMapper.getReviewList(startIdx));
+
```
페이지의 넘버가 바뀌었을때 이전페이지에서 나왔던 중복된 결과를 가져오지는 않나요?

Jeonminje replied:
currentPageNum이 바뀌는 상황을 말씀하시는 건가요??
그렇다면 제가 테스트한 결과 이전페이지에서 나왔던 중복결과를 가져오지는 않았습니다!
예를 들어 currentPageNum = 1 이라면 0 ~ 9번째, currentPageNum = 2 라면 10 ~ 19번째의 리뷰가 전달 되는 형식으로 코드를 작성 하였습니다 :)

sunwootest commented on src/main/java/study/movieservice/controller/MovieController.java:
```diff
@@ -0,0 +1,31 @@
+
+    @GetMapping("/review/list")
+    @ResponseStatus(HttpStatus.OK)
+    public String getReviewList(@RequestParam(defaultValue = "1") Integer currentPageNum){
```
defaultValue를 1로 지정할 필요가 있을까요? 어차피 클라이언트 측에서 페이지 1로 제공을 해줄 것 같은데요.

Jeonminje replied:
클라이언트 측이 없다보니 혹시 몰라 defaultValue를 지정해 놓았습니다. 혹시 수정하는 것이 좋을까요?

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -0,0 +1,45 @@
+        int totalPageNum = (int) Math.ceil((double) total/reviewPerPage);
+        int startIdx = (currentPageNum - 1) * reviewPerPage;
+
+        JSONObject jsonObject = new JSONObject();
+        jsonObject.put("total", total);
+        jsonObject.put("totalPageNum", totalPageNum);
+        jsonObject.put("currentPage", currentPageNum);
+        jsonObject.put("data", reviewMapper.getReviewList(startIdx));
```
JSONObject로 반환하기 보단 List를 활용하는 것이 나아보이네요.

Jeonminje replied:
반환 형식을 json으로 하지말라는 말씀이실까요?

Jeonminje replied:
혹시 List를 활용한다면 리뷰List 말고, 다른 정보는 리뷰list와 데이터 타입이 달라 Map을 이용한다고 해도 함께 담기 힘들 것 같은데 total, totlaPageNum과 같은 정보는 어떻게 처리할지 여쭤봐도 될까요??

nahokyun commented on src/main/java/study/movieservice/controller/MovieController.java:
```diff
@@ -0,0 +1,31 @@
+
+    @GetMapping("/review/list")
+    @ResponseStatus(HttpStatus.OK)
+    public String getReviewList(@RequestParam(defaultValue = "1") Integer currentPageNum){
+
+        return movieService.getReviewList(currentPageNum).toString();
+    }
```
ReviewList를 가져오지못하더라도 OK를 반환하는것 같습니다!

Jeonminje replied:
결과를 도출하는 것에만 집중하다보니 예외를 생각하지 못했던 거 같네요 반영해보겠습니다!

sunwootest reviewed (commented):
페이징을 잘 적용하셨네요. 기회가 된다면 페이징 쿼리 최적화 방법에 대해서도 알아보시면 좋을 것 같습니다 :)

Jeonminje committed:
## MovieController
- getReviewList()의 리턴 타입을 List로 수정하였습니다.

## ExceptionMessageConst
- 리뷰 불러오기 실패에 대한 메시지를 추가하였습니다.

## MovieService
- getReviewList()에서 JsonObject를 리턴해주던 형식에서 List를 생성하여 반환해주었습니다.
- List에 추가할 map을 만들기위해 private한 makeMap() 함수를 추가하였습니다.

Jeonminje commented:
## MovieController
- getReviewList()의 리턴 타입을 List로 수정하였습니다.

## ExceptionMessageConst
- 리뷰 불러오기 실패에 대한 메시지를 추가하였습니다.
```
FAILED_BRING_REVIEW("리뷰 불러오기에 실패했습니다.");
```
## MovieService
- getReviewList()에서 JsonObject를 리턴해주던 형식에서 List를 생성하여 반환해주었습니다. 
- List에 추가할 map을 만들기위해 private한 makeMap() 함수를 추가하였습니다.
- 만약의 상황에 대비해 try catch 로 로직을 감싸주었습니다.
```java
 public List<Map<String, Object>> getReviewList(Integer currentPageNum){

        List<Map<String, Object>> result = new ArrayList<Map<String, Object>>();

        try{
            int total = reviewMapper.getTotalRowCount();
            int reviewPerPage = 10;
            int totalPageNum = (int) Math.ceil((double) total/reviewPerPage);
            int startIdx = (currentPageNum - 1) * reviewPerPage;
            List<ReviewVO> data = reviewMapper.getReviewList(startIdx);

            result.add(makeMap("total", total));
            result.add(makeMap("totalPageNum", totalPageNum));
            result.add(makeMap("currentPage", currentPageNum));
            result.add(makeMap("data", data));
        } catch (Exception e){
            throw  new IllegalArgumentException(FAILED_BRING_REVIEW.getMessage());
        }

        return result;
    }

    private Map<String, Object> makeMap(String s, Object o){
        Map<String, Object> map = new HashMap<>();
        map.put(s, o);
        return map;
    }
```

Jeonminje committed:
Merge remote-tracking branch 'origin/develop' into Movie-review#24

# Conflicts:
#	src/main/java/study/movieservice/domain/ExceptionMessageConst.java
#	src/main/java/study/movieservice/service/MovieService.java

Jeonminje commented:
커밋했습니다!

nahokyun commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -0,0 +1,63 @@
+    private final ReviewMapper reviewMapper;
+    private final HttpSession httpSession;
+
+    public void saveReview(Review reviewDto) {
```
Review reviewDto를 파라미터로 받을때 reviewDto 안에 memberId에 해당하는 정보가 없다면 ReviewDto를 따로 만드는건 어떨까요? 불완전객체는 사용을 지양하는것이 좋다고 알고있습니다!

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -0,0 +1,63 @@
+            int startIdx = (currentPageNum - 1) * reviewPerPage;
+            List<ReviewVO> data = reviewMapper.getReviewList(startIdx);
+
+            result.add(makeMap("total", total));
+            result.add(makeMap("totalPageNum", totalPageNum));
+            result.add(makeMap("currentPage", currentPageNum));
+            result.add(makeMap("data", data));
```
key value 쌍으로 만들기 위해서 억지로 Map을 추가한 느낌이 있는데 이럴거면 total~data가 포함된 클래스를 하나 생성하는 것이 낫지 않을까요?

Jeonminje replied:
넵 감사합니다!

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -0,0 +1,63 @@
+
+        try{
+            int total = reviewMapper.getTotalRowCount();
+            int reviewPerPage = 10;
```
이런 정보들은 properties로 빼도 될 것 같네요. 스프링에서 properties를 어떻게 활용할 수 있는지 알아보면 도움이 될 것 같습니다 :) 또 더 나아가 yaml이라는 방식도 있으니 참고해보세요 :)

Jeonminje replied:
넵 알아보겠습니다.

Jeonminje committed:
변경된 develop과 merge, 페이징 로직 수정

Jeonminje commented:
>## 변경 된 점 
- 기존 리뷰 목록을 불러올 시, 컨트롤러에서 List를 이용하여 응답해주던 로직에서 PagingVO를 만들어 응답하는 것으로 수정하였습니다.
- 그에 따른 몇몇 함수의 변화가 있습니다.
- develop 브랜치의 변화로 인해 develop -> 브랜치로의 merge를 해주었습니다.

## MovieController.getReviewList()
- getReviewList()의 리턴 타입을 PagingVO로 변경하였습니다.
## PagingVO
- 페이징을 위한 클래스가 추가되었습니다.
```java
public class PagingVO {
    private Integer total;
    private Integer totalPageNum;
    private Integer currentPage;
    private Object data;
}
```
## MovieService.getReviewList()
- reviewPerPage를 프로퍼티파일에 저장함으로서 @Value 어노테이션을 사용하여 초기화 해주었습니다.
- getReviewList() List를 직접 만들던 방법에서 PagingVO를 이용하여 Builder 패턴을 이용해 데이터를 담아 주었습니다.

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -1,15 +1,71 @@
+        reviewMapper.save(review);
+    }
+
+    public PagingVO getReviewList(Integer currentPageNum){
+
+        int total;
+        int totalPageNum;
+        int startIdx = (currentPageNum - 1) * reviewPerPage;
+        List<ReviewVO> data;
+
+        try{
+            total = reviewMapper.getTotalRowCount();
+            data = reviewMapper.getReviewList(startIdx);
+            totalPageNum = (int) Math.ceil((double) total/reviewPerPage);
+        } catch (Exception e){
+            throw  new IllegalArgumentException(FAILED_BRING_REVIEW.getMessage());
+        }
+
+        PagingVO result = PagingVO.builder()
+                .total(total)
+                .totalPageNum(totalPageNum)
+                .currentPage(currentPageNum)
+                .data(data)
+                .build();
+
+        return result;
+    }
```
이제는 개발과 관련된 요구사항에 집중할 수준이 된 것 같네요. total과 data는 db에서 직접 조회하는 것을 확인할 수 있습니다. 그런데 만약에 DB에 우리가 원하는 데이터가 존재하지 않으면 그대로 NullPointerException을 던지겠죠? 이런 것들도 대비할 수 있는 메소드를 구현해야할 것 같습니다. 이런 것들이 테스트의 필요성을 요구하기도 하고요 :)

sunwootest replied:
ReviewList를 우리가 매번 DB에서 조회할 필요가 있을까요? 캐싱을 적용해볼 수도 있을 것 같은데 이런 것들은 나중에 따로 이슈로 빼서 적용하면 좋을 것 같습니다 :)

Jeonminje replied:
넵 조언 감사합니다!!

Jeonminje replied:
혹시 현재상황에서는 db에 데이터가 없다면 total은 count를 이용해 0이 대입되고, data는 List 형태라 값이 없는 List로,
값을 비교하거나 이용하는게 아닌 단순히 DB에서 가져온 데이터를 보내는 로직이라 NPE가 발생하지 않는데 이 함수에서도 그런 상황에 대비할 메소드를 구현해야 할까요??

sunwootest commented on src/main/java/study/movieservice/controller/MovieController.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final MovieService movieService;
+
+    @PostMapping("/review")
+    @ResponseStatus(HttpStatus.OK)
+    @LoginCheck
+    public String saveReview(@RequestBody Review review){
+        movieService.saveReview(review);
+        return SUCCESS_SAVE_REVIEW.getMessage();
+    }
```
컨벤션이 통일 안된 것 같은데요 Http 상태코드를 보니까. 이제 develop에 서로가 분업해서 merge 할텐데 아직도 컨벤션이 지켜지지 않으면 나중가면 유지보수가 어려워질 수 있습니다 :)

Jeonminje replied:
통일 시키겠습니다!

sunwootest commented on src/main/java/study/movieservice/service/MovieService.java:
```diff
@@ -1,15 +1,71 @@
-@RequiredArgsConstructor
 public class MovieService {
 
+    private final ReviewMapper reviewMapper;
```
개인적인 제안을 드리면 Review와 관련된 내용이 이렇게 많으면 Review 서비스로 분리해도 되지 않을까 싶습니다. 이건 참고만 하시고 필요하다 싶으면 적용해보세요 :)

Jeonminje replied:
넵 감사합니다 :)

Jeonminje committed:
ReviewService 분리

Jeonminje committed:
리뷰 페이징 수정

Jeonminje commented:
>## 변경사항
- MovieService에 있던 리뷰관련 로직을 ReviewService 를 새로 생성하여 분리하였습니다.
- 리뷰 페이징을 오래된 리뷰부터 가져오던 문제를 최근 리뷰부터 가져오도록 수정하였습니다.

Jeonminje committed:
리뷰 삭제기능 추가

Jeonminje commented:
>## 추가 변경사항
- 리뷰 삭제기능이 추가되었습니다.
```sql
    <delete id="delete">
        delete from review
        where id = #{reviewId}
    </delete>
```
recommend 테이블의 행을 함께 삭제해줘야 하기때문에 추후 recommend기능과 merge 후 비즈니스로직을 수정할 계획입니다.

Jeonminje committed:
Merge remote-tracking branch 'origin/develop' into Movie-review#24

# Conflicts:
#	src/main/java/study/movieservice/controller/MovieController.java
#	src/main/java/study/movieservice/domain/ExceptionMessageConst.java

Jeonminje committed:
merge
